graphistry/compute/chain_remote.py

```diff
@@ -0,0 +1,297 @@
+        raise ValueError("Missing dataset_id; either pass in, or call on g2=g1.plot(render='g') in api=3 mode ahead of time")
+
+    assert (engine is None) or engine in ["pandas", "cudf"], f"engine should be None, 'pandas', or 'cudf', got: {engine}" 
+
+    if format is None:
+        if output_type == "shape":
+            format = "json"
+        else:
+            format = "parquet"
+
+    if isinstance(chain, Chain):
+        chain_json = chain.to_json()
+    elif isinstance(chain, list):
+        chain_json = Chain(chain).to_json()
+    else:
+        assert isinstance(chain, dict)
+        chain_json = chain
+
+    if validate:
+        Chain.from_json(chain_json)
+
+    request_body = {
+        "gfql_operations": chain_json,
+        "format": format
+    }
+
+    if node_col_subset is not None:
+        request_body["node_col_subset"] = node_col_subset  # type: ignore
+    if edge_col_subset is not None:
+        request_body["edge_col_subset"] = edge_col_subset  # type: ignore
+    if df_export_args is not None:
+        request_body["df_export_args"] = df_export_args
+    if engine is not None:
+        request_body["engine"] = engine  # type: ignore
+
+    base_url = ""
+    url = f"{base_url}/api/v2/etl/datasets/{dataset_id}/gfql/"
+
+    # Prepare headers
+    headers = {
+        "Authorization": f"Bearer {api_token}",
+        "Content-Type": "application/json",
+    }
+
+    response = requests.post(url, headers=headers, json=request_body)
+
+    response.raise_for_status()
+
+    # deserialize based on output_type & format
+
+    if isinstance(self._edges, pd.DataFrame):
+        df_cons = pd.DataFrame
+        read_csv = pd.read_csv
+        read_parquet = pd.read_parquet
+    elif 'cudf.core.dataframe' in str(getmodule(self._edges)):
+        import cudf
+        df_cons = cudf.DataFrame
+        read_csv = cudf.read_csv
+        read_parquet = cudf.read_parquet
+    else:
+        raise ValueError(f"Unknown self._edges type, expected cudf/pandas DataFrame: {type(self._edges)}")
+
+    if output_type == "shape":
+        if format == "json":
+            return pd.DataFrame(response.json())
+        elif format == "csv":
+            return read_csv(BytesIO(response.content))
+        elif format == "parquet":
+            return read_parquet(BytesIO(response.content))
+        else:
+            raise ValueError(f"Unknown format, expected json/csv/parquet, got: {format}")
+    elif output_type == "all" and format in ["csv", "parquet"]:
+        zip_buffer = BytesIO(response.content)
+        with zipfile.ZipFile(zip_buffer, "r") as zip_ref:
+            nodes_file = [f for f in zip_ref.namelist() if "nodes" in f][0]
+            edges_file = [f for f in zip_ref.namelist() if "edges" in f][0]
+
+            nodes_data = zip_ref.read(nodes_file)
+            edges_data = zip_ref.read(edges_file)
+
+            nodes_df = read_parquet(BytesIO(nodes_data)) if format == "parquet" else read_csv(BytesIO(nodes_data))
```

Comment:
no action needed now but i imagine when we want to support > 2 binary formats we will need to refactor anywhere format string is being checked inline

Reply:
i think the bigger thing will be getting arrow downloads, i suspect that will give a noticeable speedup over parquet on many connections...